[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51252 [nnc] Don't fuse intrinsics on cpu if they are slower
* #51246 [nnc] Don't use sleef where it's slower
* #51190 [nnc] Expose vectorized math functions to jit fuser.
* #51187 [nnc] Use sleef if its symbols are available
* #51186 [nnc][trivial] Refactor llvm_jit so the wrapper class doesn't depend on ifdefs
* **#51125 [nnc] Refactor generation of intrinsics to reduce the amount of macro-hell**

The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

bertmaher committed:
[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell

The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]

bertmaher committed:
format on "[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell"


The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell"


The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell"


The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]

bertmaher committed:
Fix some compilation issues with gcc5.4, msvc; fix double dtype handling on "[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell"


The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Awesome, looks much cleaner now! I didn't check the logic assuming it stays the same.

bertmaher committed:
Update on "[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell"


The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]

bertmaher committed:
fix sleef type suffix at the appropriate point in the stack on "[nnc] Refactor generation of intrinsics to reduce the amount of macro-hell"


The big pile of X-macros used for emitting (possibly vectorized)
intrinsics makes it **really** difficult to change that code in any systematic
way (which I'm about to do in a later diff).

We can factor most of what the macro does into a fairly simple function.  There
are still macros but they're just a bunch of case/call helper/break
boilerplate.

Differential Revision: [D26078384](https://our.internmc.facebook.com/intern/diff/D26078384/)

[ghstack-poisoned]
